src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,122 @@
+import { exec } from 'child_process';
+import { ModeName, Mode } from '../../mode/mode';
+import { existsSync, exists } from 'fs';
+import { configuration } from '../../configuration/configuration';
+import * as vscode from 'vscode';
+import * as util from '../../util/util';
+import { Message } from '../../util/message';
+
+// InputMethodSwitcher change input method automatically when mode changed
+export class InputMethodSwitcher {
+  private savedIMKey = '';
+
+  public async switchInputMethod(prevMode: ModeName, newMode: ModeName) {
+    if (configuration.autoSwitchInputMethod.enable !== true) {
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    if (
+      prevMode === ModeName.Insert ||
+      prevMode === ModeName.SurroundInputMode ||
+      prevMode === ModeName.Replace
+    ) {
+      if (
+        newMode !== ModeName.Insert &&
+        newMode !== ModeName.SurroundInputMode &&
+        newMode !== ModeName.Replace
+      ) {
+        this.switchToDefaultIM();
+        return;
+      }
+    }
+    // when you enter insert-like mode, resume origin input method
+    if (
+      newMode === ModeName.Insert ||
+      newMode === ModeName.SurroundInputMode ||
+      newMode === ModeName.Replace
+    ) {
+      if (
+        prevMode !== ModeName.Insert &&
+        prevMode !== ModeName.SurroundInputMode &&
+        prevMode !== ModeName.Replace
+      ) {
+        this.resumeIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  private async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchInputMethod.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (obtainIMCmd !== '') {
+      if (existsSync(rawObtainIMCmd)) {
+        const insertIMKey = await util.executeShell(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } else {
+        this.showCmdNotFoundErrorMessage(rawObtainIMCmd, 'vim.autoSwitchInputMethod.obtainIMCmd');
+      }
+    }
+
+    const defaultIMKey = configuration.autoSwitchInputMethod.defaultIM;
+    if (defaultIMKey !== this.savedIMKey) {
+      this.switchToIM(defaultIMKey);
+    }
+  }
+
+  // resume origin inputmethod
+  private async resumeIM() {
+    if (this.savedIMKey !== configuration.autoSwitchInputMethod.defaultIM) {
+      this.switchToIM(this.savedIMKey);
+    }
+  }
+
+  private async switchToIM(imKey: string) {
+    let switchIMCmd = configuration.autoSwitchInputMethod.switchIMCmd;
+    if (!switchIMCmd.includes('{im}')) {
```

Comment:
Validation of the configuration happens in various functions. Should we consolidate them all and put them all in one spot?